### PR TITLE
Masking invalid x and/or weights in hist

### LINKED ISSUE #6483: Range determination for data with NaNs
I notice that this code will plot with warnings:

``` python
from matplotlib import pyplot as plt
from numpy import NaN
plt.hist([1, NaN, 4], range=[0,5])
```

```
/Users/trish/anaconda/lib/python2.7/site-packages/numpy/lib/function_base.py:229: RuntimeWarning: invalid value encountered in greater_equal
  keep = (tmp_a >= mn)
/Users/trish/anaconda/lib/python2.7/site-packages/numpy/lib/function_base.py:230: RuntimeWarning: invalid value encountered in less_equal
  keep &= (tmp_a <= mx)
```

Whereas this code ends up with an error because of how the range is determined:

``` python
from matplotlib import pyplot as plt
from numpy import NaN
plt.hist([1, NaN, 4])
```

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/trish/anaconda/lib/python2.7/site-packages/matplotlib/pyplot.py", line 2958, in hist
    stacked=stacked, data=data, **kwargs)
  File "/Users/trish/anaconda/lib/python2.7/site-packages/matplotlib/__init__.py", line 1812, in inner
    return func(ax, *args, **kwargs)
  File "/Users/trish/anaconda/lib/python2.7/site-packages/matplotlib/axes/_axes.py", line 6010, in hist
    m, bins = np.histogram(x[i], bins, weights=w[i], **hist_kwargs)
  File "/Users/trish/anaconda/lib/python2.7/site-packages/numpy/lib/function_base.py", line 176, in histogram
    'max must be larger than min in range parameter.')
AttributeError: max must be larger than min in range parameter.
```

If we're willing to forgive NaNs in the first case and just plot while ignoring them, wouldn't it also make sense to ignore them in the range computation?  Or if that's a bad idea, maybe we could return an error that makes it clear to the user that the problem was NaNs in the data?

The friend who showed this to me was stumped by the error message because his data was huge and he had no idea it contained NaNs.  When he saw that AttributeError he just thought "But I'm letting matplotlib decide the range, how can it be wrong?"

### PR DESCRIPTION
**Pull request by dstansby**
A rebase of #7133, fixes #6483.

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -6061,6 +6061,46 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
         bin_range = range
         del range
```
Comment (klapo) [inline lib/matplotlib/axes/_axes.py]: There is a namespace conflict here. `range` is both the range of histogram and the python function. If I do just do a quick and dirty example:
```
del range
range(np.arange(100))
```
I get the below error message:
```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-21-0eb074616bf8> in <module>()
----> 1 del range
      2 range(np.arange(100))
NameError: name 'range' is not defined
```

This seems similar to the errors generated in the tests, which comes from using `range` to refer to the built-in python function on line 6171.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -6061,6 +6061,46 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
...
 
+        def _normalize_input(inp, ename='input'):
```
Comment (dstansby) [inline lib/matplotlib/axes/_axes.py]: I think the problem is that this function has disappeared on the master branch, so I think another approach to the problem is needed.

**Comment by dstansby** [comment]
I've tidied this up a bit - using masked arrays is not working at all with the rest of the `ax.hist` function though... should I be converting them back to normal arrays with the masked values just removed?

**Comment by dopplershift** [comment]
Yeah, I think `cbook.delete_masked_points` is the usual way to go here.

**Comment by dstansby** [comment]
(this still needs a test and a rebase, but I think the code is fine now)

**Comment by efiring** [comment]
Looks like you also need to modify the docstring to note the support for masked arrays and invalid values, and also note the support for masked arrays in the whats_new entry.

**Comment by tacaswell** [comment]
Due to some on-going discussion about what to do in the case of missing data (and how the handle the averaging).

**Comment by dstansby** [comment]
Going to close this since it's been a while - I'll leave my branch around for a bit if anyone thinks it's worthwhile picking up.

**Pull request closed without merging**